### PR TITLE
feat: view count

### PR DESCRIPTION
**Pull request by jinzishuai**
Ticket: https://github.com/aitok-ai/LibreChatInternal/issues/95

﻿Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation update  
  

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration:
##

##

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jinzishuai** [comment]
# Problem: loading the shared conversation page calls the API 5 times

image

**Comment by jinzishuai** [comment]
### viewing your own convo is even worse

image

we could also be loading other convos as well.

**Comment by jinzishuai** [comment]
image

still need to retrieve the actual number from the DB.

**Comment by jinzishuai** [comment]
# Still a problem with Recommendation ViewCount

image


I think the number is wrong. We might have obtained it for the wrong conversation.

File: client/src/components/Messages/MessageHeaderButtons.jsx [hunk 1]
```diff
@@ -14,6 +14,7 @@ export default function MessageHeaderButtons() {
   const { conversationId } = conversation;
   const updateConvoMutation = useUpdateConversationMutation(conversation?.conversationId);
   const likeConvoMutation = useLikeConversationMutation(conversation?.conversationId);
+  const viewCount = conversation?.viewCount;
```
Comment (jinzishuai) [inline client/src/components/Messages/MessageHeaderButtons.jsx:17]: this should too:

```
const { viewCount } = conversation
```

File: api/server/routes/messages.js [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
 const { getMessages, likeMessage } = require('../../models/Message');
+const { increaseConvoViewCount, getSharedConvo } = require('../../models/Conversation');
 const requireJwtAuth = require('../../middleware/requireJwtAuth');
 
 router.get('/:conversationId', requireJwtAuth, async (req, res) => {
   const { conversationId } = req.params;
+  const convo = await getSharedConvo(conversationId);
+  if (convo?.user !== req.user.id) {
+    //increment view count only if the user is not the owner of the conversation
+    await increaseConvoViewCount(conversationId, req.user.id); //Added for https://github.com/aitok-ai/LibreChatInternal/issues/95
```
Comment (BlitzWrecker) [inline api/server/routes/messages.js:12]: I have found the source for all the repeated API calls. @Thomasbulls once implemented the like-a-message feature. In his [implementation](https://github.com/aitok-ai/LibreChat/blob/3b8f043e48b4f3f603a0b6d48a4c018033110a47/client/src/components/Messages/Message.jsx#L128), he makes an API call for each message in a conversation to retrieve the liked status, therefore causing the view count to increment for each message in the conversation.

Comment (jinzishuai) [reply api/server/routes/messages.js:12]: that makes a lot of sense. I've found out that the number of GET /api/messages/ call scales with the number of messages inside the conversation.

**Comment by jinzishuai** [comment]
https://github.com/aitok-ai/LibreChat/pull/142 and https://github.com/aitok-ai/LibreChat/pull/143 are  the new implementation

**Pull request closed without merging**